File: native/macos-notify/Sources/AgentBridgeNotifier/Protocol.swift
import Foundation

let maxInputBytes = 16 * 1024
let maxTitleCharacters = 256
let maxBodyCharacters = 2048
let maxIdentifierCharacters = 256
let maxDetailCharacters = 1024

enum NotificationAction: String, Codable, CaseIterable {
    case view, claim, snooze
}

enum Request {
    case post(title: String, body: String, taskID: String, actions: [NotificationAction], expiresInSeconds: Int)
    case register(activationArgv: [String])
    case unregister
    case status
    case action(action: NotificationAction, notificationID: String, taskID: String)
}

private struct PostPayload: Decodable {
    let title: String
    let body: String
    let taskID: String
    let actions: [String]
    let expiresInSeconds: Int

    enum CodingKeys: String, CodingKey {
        case title, body, actions
        case taskID = "task_id"
        case expiresInSeconds = "expires_in_seconds"
    }
}

private struct RegisterPayload: Decodable {
    let activationArgv: [String]
    enum CodingKeys: String, CodingKey { case activationArgv = "activation_argv" }
}

private struct ActionPayload: Decodable {
    let action: String
    let notificationID: String
    let taskID: String
    enum CodingKeys: String, CodingKey {
        case action
        case notificationID = "notification_id"
        case taskID = "task_id"
    }
}

struct Response: Codable {
    let ok: Bool
    let notification_id: String
    let status: String
    let detail: String

    static func posted(_ notificationID: String, _ detail: String) -> Response {
        Response(ok: true, notification_id: notificationID, status: "os_posted", detail: boundedDetail(detail))
    }

    static func registered(_ detail: String) -> Response {
        Response(ok: true, notification_id: "registration", status: "os_posted", detail: boundedDetail(detail))
    }

    static func failure(_ detail: String) -> Response {
        Response(ok: false, notification_id: "", status: "failed", detail: boundedDetail(detail))
    }
}

func parseRequest(_ data: Data) throws -> Request {
    guard data.count <= maxInputBytes else { throw ProtocolError("request exceeds 16384 bytes") }
    guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any],
          let operation = object["operation"] as? String else {
        throw ProtocolError("malformed request JSON")
    }
    func requireExact(_ fields: Set<String>) throws {
        guard Set(object.keys) == fields else { throw ProtocolError("request has unknown or missing fields") }
    }
    let decoder = JSONDecoder()
    switch operation {
    case "post":
        try requireExact(["operation", "title", "body", "task_id", "actions", "expires_in_seconds"])
        let payload: PostPayload
        do { payload = try decoder.decode(PostPayload.self, from: data) } catch { throw ProtocolError("malformed request JSON") }
        let actions = try payload.actions.map { value -> NotificationAction in
            guard let action = NotificationAction(rawValue: value) else { throw ProtocolError("invalid action") }
            return action
        }
        try boundedText(payload.title, "title", maxTitleCharacters)
        try boundedText(payload.body, "body", maxBodyCharacters)
        try opaqueID(payload.taskID, "task_id")
        guard actions == NotificationAction.allCases else { throw ProtocolError("actions must be view, claim, snooze") }
        guard (1...86_400).contains(payload.expiresInSeconds) else { throw ProtocolError("invalid expires_in_seconds") }
        return .post(title: payload.title, body: payload.body, taskID: payload.taskID, actions: actions, expiresInSeconds: payload.expiresInSeconds)
    case "register":
        try requireExact(["operation", "activation_argv"])
        let payload: RegisterPayload
        do { payload = try decoder.decode(RegisterPayload.self, from: data) } catch { throw ProtocolError("malformed request JSON") }
        try validActivationArgv(payload.activationArgv)
        return .register(activationArgv: payload.activationArgv)
    case "unregister":
        try requireExact(["operation"])
        return .unregister
    case "status":
        try requireExact(["operation"])
        return .status
    case "action":
        try requireExact(["operation", "action", "notification_id", "task_id"])
        let payload: ActionPayload
        do { payload = try decoder.decode(ActionPayload.self, from: data) } catch { throw ProtocolError("malformed request JSON") }
        guard let action = NotificationAction(rawValue: payload.action) else { throw ProtocolError("invalid action") }
        try opaqueID(payload.notificationID, "notification_id")
        try opaqueID(payload.taskID, "task_id")
        return .action(action: action, notificationID: payload.notificationID, taskID: payload.taskID)
    default:
        throw ProtocolError("unknown request operation")
    }
}

struct ProtocolError: Error, LocalizedError {
    let message: String
    init(_ message: String) { self.message = message }
    var errorDescription: String? { message }
}

func boundedText(_ value: String, _ field: String, _ maximum: Int) throws {
    guard !value.isEmpty, value.count <= maximum else { throw ProtocolError("invalid \(field)") }
}

func opaqueID(_ value: String, _ field: String) throws {
    let allowed = CharacterSet(charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789._-")
    guard !value.isEmpty, value.utf8.count <= maxIdentifierCharacters,
          value.unicodeScalars.allSatisfy({ allowed.contains($0) }) else { throw ProtocolError("invalid \(field)") }
}

func validActivationArgv(_ argv: [String]) throws {
    guard (1...16).contains(argv.count), argv.allSatisfy({ !$0.isEmpty && $0.count <= 1024 }),
          argv[0].hasPrefix("/"), !argv.contains(where: { $0.contains("\0") || $0.contains("\n") || $0.contains("\r") }) else {
        throw ProtocolError("invalid activation_argv")
    }
}

func boundedDetail(_ value: String) -> String { String(value.prefix(maxDetailCharacters)) }

File: native/windows-notify/src/protocol.rs
use serde::{Deserialize, Serialize};

pub const MAX_INPUT_BYTES: usize = 16 * 1024;
pub const MAX_TITLE_CHARS: usize = 256;
pub const MAX_BODY_CHARS: usize = 2048;
pub const MAX_IDENTIFIER_CHARS: usize = 256;
pub const MAX_DETAIL_CHARS: usize = 1024;

#[derive(Debug, Deserialize, PartialEq)]
#[serde(deny_unknown_fields, tag = "operation", rename_all = "snake_case")]
pub enum Request {
    Post {
        title: String,
        body: String,
        task_id: String,
        actions: Vec<Action>,
        expires_in_seconds: u32,
    },
    Register,
    Unregister,
    Status,
    Action {
        action: Action,
        notification_id: String,
        task_id: String,
    },
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum Action { View, Claim, Snooze }

#[derive(Debug, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Response {
    pub ok: bool,
    pub notification_id: String,
    pub status: String,
    pub detail: String,
}

impl Response {
    pub fn posted(notification_id: String, detail: impl Into<String>) -> Self {
        Self { ok: true, notification_id, status: "os_posted".to_owned(), detail: bounded_detail(detail.into()) }
    }

    pub fn registered(detail: impl Into<String>) -> Self {
        Self { ok: true, notification_id: "registration".to_owned(), status: "os_posted".to_owned(), detail: bounded_detail(detail.into()) }
    }

    pub fn failure(detail: impl Into<String>) -> Self {
        Self { ok: false, notification_id: String::new(), status: "failed".to_owned(), detail: bounded_detail(detail.into()) }
    }
}

pub fn parse_request(input: &[u8]) -> Result<Request, String> {
    if input.len() > MAX_INPUT_BYTES { return Err("request exceeds 16384 bytes".to_owned()); }
    let value: serde_json::Value = serde_json::from_slice(input).map_err(|_| "malformed request JSON".to_owned())?;
    enforce_exact_fields(&value)?;
    let request: Request = serde_json::from_value(value).map_err(|_| "malformed request JSON".to_owned())?;
    validate(&request)?;
    Ok(request)
}

fn enforce_exact_fields(value: &serde_json::Value) -> Result<(), String> {
    let object = value.as_object().ok_or_else(|| "request must be an object".to_owned())?;
    let operation = object.get("operation").and_then(serde_json::Value::as_str).ok_or_else(|| "request operation is missing".to_owned())?;
    let allowed: &[&str] = match operation {
        "post" => &["operation", "title", "body", "task_id", "actions", "expires_in_seconds"],
        "register" | "unregister" | "status" => &["operation"],
        "action" => &["operation", "action", "notification_id", "task_id"],
        _ => return Err("unknown request operation".to_owned()),
    };
    if object.len() != allowed.len() || object.keys().any(|key| !allowed.contains(&key.as_str())) {
        return Err("request has unknown or missing fields".to_owned());
    }
    Ok(())
}

fn validate(request: &Request) -> Result<(), String> {
    match request {
        Request::Post { title, body, task_id, actions, expires_in_seconds } => {
            text(title, "title", MAX_TITLE_CHARS)?;
            text(body, "body", MAX_BODY_CHARS)?;
            opaque(task_id, "task_id")?;
            if actions.as_slice() != [Action::View, Action::Claim, Action::Snooze] { return Err("actions must be view, claim, snooze".to_owned()); }
            if *expires_in_seconds == 0 || *expires_in_seconds > 86_400 { return Err("invalid expires_in_seconds".to_owned()); }
        }
        Request::Action { action: _, notification_id, task_id } => {
            opaque(notification_id, "notification_id")?;
            opaque(task_id, "task_id")?;
        }
        Request::Register | Request::Unregister | Request::Status => {}
    }
    Ok(())
}

fn text(value: &str, field: &str, maximum: usize) -> Result<(), String> {
    if value.is_empty() || value.chars().count() > maximum { Err(format!("invalid {}", field)) } else { Ok(()) }
}

fn opaque(value: &str, field: &str) -> Result<(), String> {
    if value.is_empty() || value.len() > MAX_IDENTIFIER_CHARS || !value.bytes().all(|b| b.is_ascii_alphanumeric() || matches!(b, b'.' | b'_' | b'-')) { Err(format!("invalid {}", field)) } else { Ok(()) }
}

fn bounded_detail(mut detail: String) -> String {
    if detail.chars().count() > MAX_DETAIL_CHARS { detail = detail.chars().take(MAX_DETAIL_CHARS).collect(); }
    detail
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn accepts_only_the_fixed_post_shape() {
        let request = parse_request(br#"{"operation":"post","title":"T","body":"B","task_id":"opaque-1","actions":["view","claim","snooze"],"expires_in_seconds":30}"#).unwrap();
        assert!(matches!(request, Request::Post { .. }));
    }

    #[test]
    fn rejects_unknown_fields_and_executable_action_data() {
        assert!(parse_request(br#"{"operation":"register","command":"calc.exe"}"#).is_err());
        assert!(parse_request(br#"{"operation":"post","title":"T","body":"B","task_id":"opaque-1","actions":["view"],"expires_in_seconds":30}"#).is_err());
    }

    #[test]
    fn rejects_oversized_input() {
        assert!(parse_request(&vec![b' '; MAX_INPUT_BYTES + 1]).is_err());
    }
}
